CreatedAt and UpdatedAt fields on wallet records

tplooker opened the pull request:
#### Short description of what this resolves:

Adds support for the CreatedAt and UpdatedAt fields on wallet records.

#### Changes proposed in this pull request:

- Added CreatedAt and UpdatedAt to RecordBase with get only accessors
- Added support in the wallet record service for updating these fields automatically 
- Added related test cases

**Fixes**: #74

tmarkovski commented:
I have some doubts on using date format with these values. I'd like to get your thoughts on this.
Created and modified attributes should be searchable. A very common use case would be "show my credentials issued in the last week/month/year". 
We should be able to use search query like `SearchQuery.Greater("CreatedAt", DateTime.Now.AddYears(-1).ToString("O"));`
The formatted date as it is currently, isn't searchable, since it's treated as a string. Now I'm not entirely certain if string comparison would still succeed because of the way the format is year-month-date, it actually may end up returning correct results, but the timezone comparisson would fail.
Adding few unit tests that use the search API would be very beneficial.

What do you think about using Unix epoch time instead? According to the WQL specs, numbered values should be able to be searched using this API. 
The revocation model uses this format both for Indy SDK and stored in the ledger, it seems more consistent to use this value. While it's not the most common use of date types in C#, it's also not unusual, since timestamped integer values can be found everywhere. That said, do you think we should explore using `long` instead and save the values in UTC time`DateTimeOffset.UtcNow.ToUnixTimeSeconds()`? We can add couple of extension methods to `GetCreatedLocalDateTime()` or similar. Additionally, we should add some comments on these methods that they are searchable with a one liner example, similar to the code snippets that come in Indy SDK.
Thoughts?

tmarkovski commented on src/AgentFramework.Core/Models/Records/RecordBase.cs:
```diff
@@ -10,15 +10,48 @@ namespace AgentFramework.Core.Models.Records
+                    return _instantiatedAt;
+                return DateTime.Parse(createdAtStr);
+            }
+        }
```
It's OK to have `internal` setter here and use that to set the value in record base, since you can apply special logic that's visible on the property.

tplooker commented:
@tmarkovski I have found the issue, if a tag is stored encrypted then we cannot search using predicates, instead only exact matching. See here https://github.com/hyperledger/indy-sdk/tree/master/doc/design/011-wallet-query-language I will work on a fix

tplooker commented:
@tmarkovski I have now fixed and implemented this, ideally down the track I would like to setup an attribute we can use to decorate wallet record properties to indicate whether they are searchable with non-equality type query predicates

tmarkovski commented on src/AgentFramework.Core/Contracts/IDateTimeHelper.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace AgentFramework.Core.Contracts
+{
+    public interface IDateTimeHelper
```
This helper is not necessary for simple implementations.

tmarkovski commented:
> @tmarkovski I have found the issue, if a tag is stored encrypted then we cannot search using predicates, instead only exact matching. See here https://github.com/hyperledger/indy-sdk/tree/master/doc/design/011-wallet-query-language I will work on a fix

Great find. We'll do some hardening on this later.
